### ISSUE TITLE
[question] Is there any way to prevent the mpv window from stealing focus when it pops up during playback?

### ISSUE THREAD

**Issue by SexySteak** [rany2/edge-tts#62]
I use "edge-playback" to started mpv and I have set "window-minimized" in the "mpv.conf" beforehand. mpv correctly started minimized, but it still took focus. 
You can see your cmd window (or your vscode windows if you use the cli of vscode) lose the focus clearly. After the mpv playback completed and automatically closed, the cmd window regain the focus.
Is there any solutions to fix it? I have searched Issues of MPV and found that the similar problem has been reported for years, but still not be fixed. Or perhaps there are any other ways use playback?

**Comment by LordFrz** [comment]
`subprocess.run(["mpv", "--ontop=no", "--video=no", "--title=TTS VOICE", "--screen=1", "--window-minimized=yes", OUTPUT_FILE], creationflags=subprocess.CREATE_NO_WINDOW)`

is what I'm using, I had the same issues and looked everywhere. The main thing you need to add is the bit on the end.  The creation flag will cause it not to steal focus. Not sure how that will look in the conf file but should be similar. 

I could not find the answer anywhere, but chatGPT spat it out instantly, lol

**Comment by LordFrz** [comment]
if you need to have it use the conf file instead of directly in code, then you can try these.
`no-osd = yes`
and if that one still doesn't work try adding this one too.
`no-input-default-bindings = yes`
This option disables all default input bindings for MPV, which can prevent it from responding to keyboard and mouse input when it launches.

Disabling default input bindings will prevent you from using certain keyboard shortcuts or mouse gestures in MPV. If you want to use custom input bindings, you can define them in your configuration file using the input.conf option.

**Comment by SexySteak** [comment]
> `subprocess.run(["mpv", "--ontop=no", "--video=no", "--title=TTS VOICE", "--screen=1", "--window-minimized=yes", OUTPUT_FILE], creationflags=subprocess.CREATE_NO_WINDOW)`
> 
> is what I'm using, I had the same issues and looked everywhere. The main thing you need to add is the bit on the end. The creation flag will cause it not to steal focus. Not sure how that will look in the conf file but should be similar.
> 
> I could not find the answer anywhere, but chatGPT spat it out instantly, lol

I have tried it but it didn't work. I checked the source code of mpv and found that it uses `ShowWindow(w32->window, SW_SHOWMINIMIZED);` instead of ` ShowWindow(w32->window, SW_SHOWMINNOACTIVE);`.
SW_SHOWMINIMIZED will minimize the window at startup but capture focus. I have submitted a PR for a simple fix, but no one has attended to it for now. However, I found that VLC does not have this problem. Using VLC, adjusting the settings to start minimized will not cause this issue!
By the way, it would be great if edge-tts had a better playback logic. Similar to when the media playback is finished, the player waits in the background instead of closing, and Python can still run normally, so the next time the media is played, the old process can be reused. I think this may require more complex inter-process communication.

**Comment by LordFrz** [comment]
the `, creationflags=subprocess.CREATE_NO_WINDOW` can be added to line 63 of the __main__.py of the edge_playback code, and works for me. However you still cant use the command console until its does because its waiting, but anything you type will showup after its done. ANd the command window keeps focus throughout. 

Best of luck

**Comment by person2217** [comment]
The chatgpt generated answers are good, I am using：
Configuring GoldenDict for Custom Text-to-Speech (TTS) with Python Scripts
This guide explains how to set up GoldenDict to use Python scripts for Text-to-Speech (TTS) in different languages. We will configure GoldenDict to execute custom Python scripts to generate and play audio for selected words or phrases.

**Requirements**

1. Python 3 installed on your system.
2. edge-tts module (Text-to-Speech service) installed：
`pip install edge-tts`
3. mpv media player installed for playing the audio in the background.
4. Custom Python scripts for each language (examples provided below).

 **Step 1: Python Script Setu**p

Create Python scripts for each language you want to use in GoldenDict. These scripts will generate the TTS audio and play it using mpv without bringing up a visible window.

Example Python Script for English (play_tts_en.py)

```
import os
import subprocess
import sys

# Get the input text from command-line arguments
text = sys.argv[1] if len(sys.argv) > 1 else "Hello, world!"
output_file = "output_audio.mp3"  # Output audio file

# Generate TTS audio using edge-tts
subprocess.run(["edge-tts", "--voice", "en-US-JennyNeural", "--text", text, "--write-media", output_file])

# Play the generated audio using mpv
subprocess.run([
    "mpv", 
    "--no-terminal", 
    "--really-quiet", 
    "--ontop=no", 
    "--video=no", 
    "--title=TTS VOICE", 
    "--screen=1", 
    "--window-minimized=yes", 
    output_file
], creationflags=subprocess.CREATE_NO_WINDOW)

```
Here, change **"en-US-JennyNeural"** to the voice you need.`edge-tts --list-voices` query
Create similar scripts for other languages by changing the --voice argument.

**Step 2: Configuring GoldenDict**

1. Open GoldenDict and navigate to Edit > Preferences.
2. Go to the Audio tab where you can add custom TTS engines.
3. Add a new engine for each language and set the appropriate Python script. Use the following format:
**English TTS**
`python D:/path_to_scripts/play_tts_en.py "%GDWORD%"`
Make sure to replace `D:/path_to_scripts/` with the actual path where your scripts are located.

**Step 3: Testing**
Once the configuration is done, simply look up a word in GoldenDict, and it should execute the Python script to generate the TTS audio and play it. The audio file will be deleted after playback.

**Troubleshooting**
Ensure that mpv is installed and accessible in your system's PATH.
Verify that python is correctly installed and can run from the command line.
If the script does not execute, check for Python errors by running the command directly from a terminal to see any error output.

**Comment by yjg30737** [comment]
I've just made whole gui which can solve the problem:

https://github.com/yjg30737/edge-tts-gui

TL;DR

```
with subprocess.Popen(
                [
                    "mpv",
                    filename,
                ], shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
        ) as proc:
    proc.communicate()
```

This might work.
